p2p/src/cybershake.rs

```diff
@@ -218,63 +239,134 @@ where
+        LittleEndian::write_u16(&mut self.buf[..2], buf_len);
+        self.buf.as_mut_slice()[2..TEXT_START_POSITION].copy_from_slice(tag.as_slice());
 
         self.seq += 1;
+        self.flushing = true;
+    }
 
-        // Write the length prefix and the ciphertext.
-        self.writer
-            .write(&encode_u64le(ciphertext.len() as u64)[..])
-            .await?;
-        self.writer.write(&ciphertext[..]).await?;
-        self.writer.flush().await?;
-        Ok(())
+    fn flush_pending_ciphertext(&mut self, cx: &mut Context) -> Poll<Result<(), io::Error>> {
+        if !self.flushing {
+            return Poll::Ready(Ok(()));
+        }
+        while self.ciphertext_sent < self.buf.len() {
+            let poll = self
+                .writer
+                .as_mut()
+                .poll_write(cx, &self.buf[self.ciphertext_sent..]);
+            let n = ready!(poll);
+            self.ciphertext_sent += n;
+        }
+        ready!(self.writer.as_mut().poll_flush(cx));
+        self.ciphertext_sent = 0;
+        self.flushing = false;
+        self.buf.truncate(TEXT_START_POSITION);
+        Poll::Ready(Ok(()))
     }
 }
 
-impl<W: AsyncRead + Unpin> Incoming<W> {
-    pub async fn receive_message(&mut self) -> Result<Vec<u8>, Error> {
-        let mut lenbuf = [0u8; 8];
-        let seq = self.seq;
-        self.seq += 1;
-        self.reader.read_exact(&mut lenbuf[..]).await?;
-        let len = LittleEndian::read_u64(&lenbuf) as usize;
-
-        // length must include IV prefix (16 bytes)
-        if len < 16 {
-            return Err(Error::ProtocolError);
-        }
-        // Check the message length and fail before changing any of the remaining state.
-        if (len - 16) > self.message_maxlen {
-            return Err(Error::MessageTooLong(len - 16));
+impl<W: AsyncWrite + Unpin> AsyncWrite for Outgoing<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let me = self.get_mut();
+
+        ready!(me.flush_pending_ciphertext(cx));
+
+        if me.buf.len() + buf.len() >= PLAINTEXT_BUF_SIZE as usize + 2 {
+            // plaintext_buf has BUF_SIZE size, so subtract with overflow will be never.
+            let size_to_write = PLAINTEXT_BUF_SIZE as usize + 2 - me.buf.len();
+            me.buf.extend_from_slice(&buf[..size_to_write]);
+            me.cipher_buf();
+            ready!(me.flush_pending_ciphertext(cx));
+            Poll::Ready(Ok(size_to_write))
+        } else {
+            me.buf.extend_from_slice(buf);
+            Poll::Ready(Ok(buf.len()))
         }
+    }
 
-        let mut ciphertext = Vec::with_capacity(len);
-        ciphertext.resize(len, 0u8);
-        self.reader.read_exact(&mut ciphertext[..]).await?;
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        let me = self.get_mut();
+        if me.buf.len() == 0 {
```

Comment:
i think that's a bug. WriteZero error is intended when we try to write >0, but got Ok(0) from the underlying writer. But if we are flushing an empty buffer, that should be a no-op.

Reply:
Changed returning value from Err to Ok